spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java

```diff
@@ -0,0 +1,156 @@
+package org.springframework.cloud.sleuth.instrument.rxjava2;
+
+import java.lang.reflect.Field;
+import java.util.List;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.cloud.sleuth.Span;
+import org.springframework.cloud.sleuth.TraceKeys;
+import org.springframework.cloud.sleuth.Tracer;
+
+import io.reactivex.disposables.Disposable;
+import io.reactivex.functions.Function;
+import io.reactivex.plugins.RxJavaPlugins;
+
+/**
+ * {@link RxJavaPlugins} setup schedule handler into tracing for all schedulers
+ * representation.
+ *
+ * @author Łukasz Guz
+ * @author Jakub Pyda
+ */
+class SleuthRxJava2SchedulersHandler {
+
+	private static final Log log = LogFactory
+			.getLog(SleuthRxJava2SchedulersHandler.class);
+
+	private static final String RXJAVA_COMPONENT = "rxjava2";
+
+	SleuthRxJava2SchedulersHandler(Tracer tracer, TraceKeys traceKeys,
+			List<String> threadsToSample) {
+		try {
+			Function<? super Runnable, ? extends Runnable> delegate = RxJavaPlugins
+					.getScheduleHandler();
+			if (delegate instanceof SleuthRxJava2SchedulersHandler.ScheduleHandler) {
+				return;
+			}
+			logCurrentStateOfRxJavaPlugins();
+			RxJavaPlugins.reset();
+			RxJavaPlugins.setScheduleHandler(
+					new ScheduleHandler(tracer, traceKeys, threadsToSample, delegate));
+		}
+		catch (Exception e) {
+			log.error("Failed to register Sleuth RxJava SchedulersHook", e);
+		}
+	}
+
+	private void logCurrentStateOfRxJavaPlugins() {
+		if (log.isDebugEnabled()) {
+			log.debug("Registering Sleuth RxJava2 Schedulers Hook.");
+		}
+	}
+
+	static class ScheduleHandler implements Function<Runnable, Runnable> {
+
+		private final Tracer tracer;
+		private final TraceKeys traceKeys;
+		private final List<String> threadsToSample;
+		private final Function<? super Runnable, ? extends Runnable> delegate;
+
+		public ScheduleHandler(Tracer tracer, TraceKeys traceKeys,
+				List<String> threadsToSample,
+				Function<? super Runnable, ? extends Runnable> delegate) {
+			this.tracer = tracer;
+			this.traceKeys = traceKeys;
+			this.threadsToSample = threadsToSample;
+			this.delegate = delegate;
+		}
+
+		@Override
+		public Runnable apply(Runnable action) throws Exception {
+			if (isTraceActionDecoratedByRxWorker(action)) {
+				return action;
+			}
+			Runnable wrappedAction = this.delegate != null ? this.delegate.apply(action)
+					: action;
+			return new SleuthRxJava2SchedulersHandler.TraceAction(this.tracer,
+					this.traceKeys, wrappedAction, this.threadsToSample);
+		}
+
+		private boolean isTraceActionDecoratedByRxWorker(Runnable action) {
+			try {
+				if (action instanceof Disposable) {
```

Comment:
this is likely an expensive check. maybe check w/ some rx maintainers to see if there's a better way?